### PR TITLE
Chainer: Simplify n_step_rnn test

### PR DESCRIPTION
**Pull request by dido1998**

### CODE REVIEW

File: tests/chainer_tests/functions_tests/rnn_tests/test_function_n_step_rnn.py [hunk 1]
```diff
@@ -4,569 +4,257 @@
...
+    # CPU tests
+    + testing.product({
+        'use_cuda': [False],
+        'use_ideep': ['never', 'always'],
+    })
+    # GPU tests
+    + testing.product([
+        [{'use_cuda': True}],
+
+        testing.product({
```
Comment (niboshi) [inline tests/chainer_tests/functions_tests/rnn_tests/test_function_n_step_rnn.py]: Please add `[{use_cudnn': 'never'}] +`

File: tests/chainer_tests/functions_tests/rnn_tests/test_function_n_step_rnn.py [hunk 2]
```diff
@@ -4,569 +4,257 @@
 
 import chainer
 from chainer.backends import cuda
-from chainer import functions
+import chainer.functions as F
+import chainer.functions as functions
```
Comment (niboshi) [inline tests/chainer_tests/functions_tests/rnn_tests/test_function_n_step_rnn.py]: `functions` is not needed?

File: tests/chainer_tests/functions_tests/rnn_tests/test_function_n_step_rnn.py [hunk 3]
```diff
@@ -4,569 +4,257 @@
...
 from chainer import gradient_check
 from chainer import testing
+from chainer import Variable
```
Comment (niboshi) [inline tests/chainer_tests/functions_tests/rnn_tests/test_function_n_step_rnn.py]: Please avoid importing a class directly, as per the OpenStack coding guideline.

File: tests/chainer_tests/functions_tests/rnn_tests/test_function_n_step_rnn.py [hunk 4]
```diff
@@ -4,569 +4,257 @@
...
+        out = F.n_step_rnn(self.n_layers, 0.0, h, ws, bs, xs)
+        rets = []
+        rets.append(out[0])
+        for i in range(len(out[1])):
+            rets.append(out[1][i])
+        return tuple(rets)
+
+    def forward_expected(self, inputs):
+        h, ws, bs, xs = self.process_inputs(inputs)
+        with chainer.using_config('use_ideep', 'never'):
```
Comment (niboshi) [inline tests/chainer_tests/functions_tests/rnn_tests/test_function_n_step_rnn.py]: Is this needed?

File: tests/chainer_tests/functions_tests/rnn_tests/test_function_n_step_rnn.py [hunk 5]
```diff
@@ -4,569 +4,257 @@
...
+            index += 4
+
+        return h, ws, bs, xs
+
+    def forward(self, inputs, device):
+        h, ws, bs, xs = self.process_inputs(inputs)
+        # For some reason even though only float32 arrays are created in generate_inputs(),
+        # arrays coming as input here are of type float32 and float64
+        if h.array.dtype == numpy.float64:
+            raise unittest.SkipTest('float64 not supported')
```
Comment (niboshi) [inline tests/chainer_tests/functions_tests/rnn_tests/test_function_n_step_rnn.py]: I guess this happens when calculating numeric gradient (`gradient_check.numeric_grad` casts the inputs to `float64` for better precision). So how about casting back to `float32` here?

**Comment by niboshi** [comment]
Thanks! Please check the review.

File: tests/chainer_tests/functions_tests/rnn_tests/test_function_n_step_rnn.py (hunk 4 above)
Comment (dido1998) [reply tests/chainer_tests/functions_tests/rnn_tests/test_function_n_step_rnn.py]: Actually, I was following `test_convolution_2d.py` (https://github.com/chainer/chainer/blob/e2e6eaa90f4f01d905767b69503d56629b350055/tests/chainer_tests/functions_tests/connection_tests/test_convolution_2d.py#L110), that's why I had added this. I will remove it.

**Comment by dido1998** [comment]
I cast it back to `float32` as you suggested, but the tests just halt on my local machine after a few seconds. They were not progressing further even after waiting for a long time. I have never encountered such a scenario before. Do you know why this maybe happening?

**Comment by niboshi** [comment]
I couldn't replicate that.
Does that occur no matter how many times you retry the test?
What's the test in question, specifically, and at which line of code?

**Comment by dido1998** [comment]
Now it is not occurring, I fixed it by disabling cudnn for `float64` types.

File: tests/chainer_tests/functions_tests/rnn_tests/test_function_n_step_rnn.py [hunk 6]
```diff
@@ -23,77 +14,127 @@ def _relu(x):
...
+            ws.append(inputs[index: index + 2])
+            bs.append(inputs[index + 2: index + 4])
+            index += 4
+
+        return h, ws, bs, xs
+
+    def forward(self, inputs, device):
+        h, ws, bs, xs = self.process_inputs(inputs)
+        if h.array.dtype == numpy.float64:
+            with chainer.using_config('use_cudnn', 'never'):
```
Comment (niboshi) [inline tests/chainer_tests/functions_tests/rnn_tests/test_function_n_step_rnn.py]: The previous test code did not have this workaround.
The problem is that `FunctionTestCase` passes `dtype=float64` to `gradient_check.check_backward()` and there's no way to change this behavior.
I sent a PR to make it configurable: #7817
After #7817 merged, lets avoid this workaround and set `numerical_grad_dtype = None`.

Comment (dido1998) [reply tests/chainer_tests/functions_tests/rnn_tests/test_function_n_step_rnn.py]: Yes, sure!

**Comment by emcastillo** [comment]
#7817 merged

File: tests/chainer_tests/functions_tests/rnn_tests/test_function_n_step_rnn.py (hunk 6 above)
Comment (niboshi) [reply tests/chainer_tests/functions_tests/rnn_tests/test_function_n_step_rnn.py]: #7817 has been merged!

**Comment by dido1998** [comment]
@niboshi, I have added `numerical_grad_type = None` but some backward tests fail and the tests just stop after running for a few seconds. I don't know why they are just stopping in the middle.

**Comment by niboshi** [comment]
@dido1998 First, please fix flake8 errors and numerical errors by configuring tolerance (`rtol` and `atol`) correctly.
Let's fix timeout issue later. Perhaps it's because of increased parametrization.

**Comment by dido1998** [comment]
I have made the changes, could you have a look?

**Comment by niboshi** [comment]
>tests/chainer_tests/functions_tests/rnn_tests/test_function_n_step_rnn.py::TestNStepRNN_use_chainerx_false__chainerx_device_None__use_cuda_false__cuda_device_None__use_cudnn_never__cudnn_deterministic_false__autotune_false__cudnn_fast_batch_normalization_false__use_ideep_never_param_7_{n_layers=5, hidden_size=10, input_size=6, batches=(6, 5, 3), activation='tanh'}::test_backward

This test takes quite a time.
Why is this pattern needed? Isn't the following just enough?
>```
>       {'n_layers': 2, 'hidden_size': 2, 'input_size': 3,
>            'batches': (3, 2, 1), 'activation': 'tanh'},
>```

**Comment by dido1998** [comment]
yes, you're right, it was because of the last 2 tests. Now I have removed those tests and it is running fine locally.

File: tests/chainer_tests/functions_tests/rnn_tests/test_function_n_step_rnn.py [hunk 7]
```diff
@@ -105,261 +137,130 @@ def check_forward(
...
+            ws.append(inputs[index + 4: index + 6])
+            bs.append(inputs[index + 6: index + 8])
+            index += 8
+        return h, ws, bs, xs
+
+    def forward(self, inputs, device):
+        h, ws, bs, xs = self.process_inputs(inputs)
+
+        if h.array.dtype == numpy.float64:
+            with chainer.using_config('use_cudnn', 'never'):
```
Comment (niboshi) [inline tests/chainer_tests/functions_tests/rnn_tests/test_function_n_step_rnn.py]: Now with `numerical_grad_dtype = None`, I think this workaround can be removed too,

Comment (dido1998) [reply tests/chainer_tests/functions_tests/rnn_tests/test_function_n_step_rnn.py]: Done

**Comment by niboshi** [comment]
Thanks.
Jenkins, test this please

**Pull request merged**